server/permalinks_test.go

```diff
@@ -0,0 +1,421 @@
+			replacements: []replacement{
+				{
+					index: 6,
+					word:  "https://gitlab.com/mattermost/mattermost-server/-/blob/cbb25838a61872b624ac512556d7bc932486a64c/app/authentication.go#L15-L22",
+					permalinkInfo: struct {
+						haswww string
+						commit string
+						user   string
+						repo   string
+						path   string
+						line   string
+					}{
+						commit: "cbb25838a61872b624ac512556d7bc932486a64c",
+						haswww: "",
+						line:   "L15-L22",
+						path:   "app/authentication.go",
+						user:   "mattermost",
+						repo:   "mattermost-server",
+					},
+				}, {
+					index: 142,
+					word:  "https://gitlab.com/mattermost/mattermost-server/-/blob/cbb25838a61872b624ac512556d7bc932486a64c/app/authentication.go#L15-L22",
+					permalinkInfo: struct {
+						haswww string
+						commit string
+						user   string
+						repo   string
+						path   string
+						line   string
+					}{
+						commit: "cbb25838a61872b624ac512556d7bc932486a64c",
+						haswww: "",
+						line:   "L15-L22",
+						path:   "app/authentication.go",
+						user:   "mattermost",
+						repo:   "mattermost-server",
+					},
+				},
+			},
+		},
+		{
+			name:   "bad commit hash",
+			input:  "start https://gitlab.com/mattermost/mattermost-server/-/blob/badhash/app/authentication.go#L15-L22 lorem ipsum",
+			output: "start https://gitlab.com/mattermost/mattermost-server/-/blob/badhash/app/authentication.go#L15-L22 lorem ipsum",
+			replacements: []replacement{
+				{
+					index: 6,
+					word:  "https://gitlab.com/mattermost/mattermost-server/-/blob/badhash/app/authentication.go#L15-L22",
+					permalinkInfo: struct {
+						haswww string
+						commit string
+						user   string
+						repo   string
+						path   string
+						line   string
+					}{
+						haswww: "",
+						commit: "badhash",
+						line:   "L15-L22",
+						path:   "app/authentication.go",
+						user:   "mattermost",
+						repo:   "mattermost-server",
+					},
+				},
+			},
+		},
+		{
+			name:   "bad line range",
+			input:  "start https://gitlab.com/mattermost/mattermost-server/-/blob/cbb25838a61872b624ac512556d7bc932486a64c/app/authentication.go#L22-L15 lorem ipsum",
+			output: "start https://gitlab.com/mattermost/mattermost-server/-/blob/cbb25838a61872b624ac512556d7bc932486a64c/app/authentication.go#L22-L15 lorem ipsum",
+			replacements: []replacement{
+				{
+					index: 6,
+					word:  "https://gitlab.com/mattermost/mattermost-server/-/blob/cbb25838a61872b624ac512556d7bc932486a64c/app/authentication.go#L22-L15",
+					permalinkInfo: struct {
+						haswww string
+						commit string
+						user   string
+						repo   string
+						path   string
+						line   string
+					}{
+						haswww: "",
+						commit: "cbb25838a61872b624ac512556d7bc932486a64c",
+						line:   "L22-L15",
+						path:   "app/authentication.go",
+						user:   "mattermost",
+						repo:   "mattermost-server",
+					},
+				},
+			},
+		},
+	}
```

Comment:
Add one more test for "bad file content".

Reply:
Discussed.